### DISCUSSION TITLE
Error [ERR_REQUIRE_ESM] when working with FormData

### DISCUSSION THREAD

**Discussion by PhaedrusTheGreek** [node-fetch/node-fetch#1417] [Q&A]
Trying to find a working NodeJs combination for POSTing a `multipart/form-data` , and as mentioned in the docs, I tried a few options including `formdata-polyfill` and ` formdata-node`, and also by using `(await new Response(new URLSearchParams()).formData()).constructor`, In all 3 cases I was unable to run the code, with the following error (or similar):

```
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /path/to/../node_modules/node-fetch/src/index.js
require() of ES modules is not supported.
```

I am getting that error on both v2 and v3.
Here is what i have tried:

```js
import { fileFromSync } from 'fetch-blob/from.js';
import { FormData } from 'formdata-polyfill/esm.min.js';

const fd = new FormData();
fd.append('file', blob);

const stats = fs.statSync(savedObjects);
const fileSizeInBytes = stats.size;

const response = await fetch(
  `${config.endpoint}/api/_import`,
  {
    method: 'POST',
    headers: {
      Authorization: `...`,
      'Content-Type': 'multipart/form-data',
      'Content-length': `${fileSizeInBytes}`
    },
    body: fd,
  }
);
```

**Comment by jimmywarting** [comment]
Hmm, you are not suppose to add the content-length in manually. 
The multipart payload adds extra bytes around in the body like file headers and boundary sections.

![image](https://user-images.githubusercontent.com/1148376/146089327-0f786efb-73cd-433b-a0af-23f84e89a095.png)

The byte length will be automatically calculated for you when we iterating over all the formdata fields, so you shouldn't add the content-length yourself

also, you shouldn't add the content-type manually like this cuz it needs a boundary meta data in the header in order to properly parse it. like `content-type: "multipart/form-data; boundary=----WebKitFormBoundary1qciPRkyVztmiunJ"`
...again this is also something the fetch spec adds in automatically for you

---

here is an working example: 

```js
import fetch from 'node-fetch'
import { fileFromSync } from 'fetch-blob/from.js'
import { FormData } from 'formdata-polyfill/esm.min.js'

const file = fileFromSync('./fetch-issue-1416/index.js')
const fd = new FormData()
fd.append('file', file)
fd.set('other data', 'some data')

const res = await fetch(`https://httpbin.org/post`, {
  method: 'POST',
  headers: {
    Authorization: `abc123`
  },
  body: fd
})

const json = await res.json()
console.log(json)
```

with the output being: 
```js
{
  args: {},
  data: '',
  files: {
    file: "import fetch from 'node-fetch'\n" +
      "import { fileFromSync } from 'fetch-blob/from.js'\n" +
      "import { FormData } from 'formdata-polyfill/esm.min.js'\n" +
      '\n' +
      "const file = fileFromSync('./fetch-issue-1416/index.js')\n" +
      'const fd = new FormData()\n' +
      "fd.append('file', file)\n" +
      "fd.set('other data', 'some data')\n" +
      '\n' +
      'const res = await fetch(`https://httpbin.org/post`, {\n' +
      "  method: 'POST',\n" +
      '  headers: {\n' +
      '    Authorization: `abc123`\n' +
      '  },\n' +
      '  body: fd\n' +
      '})\n' +
      '\n' +
      'const json = await res.json()\n' +
      'console.log(json)'
  },
  form: { 'other data': 'some data' },
  headers: {
    Accept: '*/*',
    'Accept-Encoding': 'gzip,deflate,br',
    Authorization: 'abc123',
    'Content-Length': '742',
    'Content-Type': 'multipart/form-data; boundary=----5570292816803601957728722138',
    Host: 'httpbin.org',
    'User-Agent': 'node-fetch'
  },
  json: null,
  origin: 'xx.xxx.xx.xx',
  url: 'https://httpbin.org/post'
}
```

**Comment by jimmywarting** [comment]
Plus the error message you gave is not about formdata, it's about ESM

to use esm, you basically need to add `"type": "module"` in the package.json or use dynamic import if you would like to use it from cjs like this:

```js
import('node-fetch').then({ default: fetch, Headers, Response, Request } => {
  fetch(url).then(...)
})
```

more about this in https://github.com/node-fetch/node-fetch/issues/1279
